adafruit_connection_manager.py

```diff
@@ -0,0 +1,362 @@
+
+    def close_socket(self, socket: SocketType) -> None:
+        """Close a previously opened socket."""
+        logger.debug("ConnectionManager.close_socket()")
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not managed")
+        key = self._get_key_for_socket(socket)
+        logger.debug(f"  closing {key}")
+        socket.close()
+        del self._socket_free[socket]
+        del self._open_sockets[key]
+
+    def free_socket(self, socket: SocketType) -> None:
+        """Mark a previously opened socket as free so it can be reused if needed."""
+        logger.debug("ConnectionManager.free_socket()")
+        if socket not in self._open_sockets.values():
+            raise RuntimeError("Socket not managed")
+        key = self._get_key_for_socket(socket)
+        logger.debug(f"  flagging {key} as free")
+        self._socket_free[socket] = True
+
+    # pylint: disable=too-many-locals,too-many-statements
+    def get_socket(
+        self,
+        host: str,
+        port: int,
+        proto: str,
+        *,
+        timeout: float = 1,
+        is_ssl: bool = False,
+        ssl_context: Optional[SSLContextType] = None,
+        max_retries: int = 5,
+        exception_passthrough: bool = False,
+    ) -> CircuitPythonSocketType:
+        """Get a new socket and connect"""
+        # pylint: disable=too-many-branches
+        logger.debug("ConnectionManager.get_socket()")
+        logger.debug(f"  tracking {len(self._open_sockets)} sockets")
+        key = (host, port, proto)
+        logger.debug(f"  getting socket for {key}")
+        if key in self._open_sockets:
+            socket = self._open_sockets[key]
+            if self._socket_free[socket]:
+                logger.debug(f"  found existing {key}")
+                self._socket_free[socket] = False
+                return socket
+
+        if proto == "https:":
+            is_ssl = True
+        if is_ssl and not ssl_context:
+            raise RuntimeError(
+                "ssl_context must be set before using adafruit_requests for https"
+            )
+
+        addr_info = self._socket_pool.getaddrinfo(
+            host, port, 0, self._socket_pool.SOCK_STREAM
+        )[0]
+
+        retry_count = 0
+        socket = None
+        last_exc = None
+        last_exc_new_type = None
+        while retry_count < max_retries and socket is None:
+            if retry_count > 0:
+                logger.debug(f"  retry #{retry_count}")
+                if any(self._socket_free.items()):
+                    self._free_sockets()
+                else:
+                    raise RuntimeError("Sending request failed") from last_exc
+            retry_count += 1
+
+            try:
+                socket = self._socket_pool.socket(addr_info[0], addr_info[1])
+            except OSError as exc:
+                logger.debug(f"  OSError getting socket: {exc}")
+                last_exc_new_type = SocketGetOSError
```

Comment:
These add support so `MiniMQTT` can handle retries

Reply:
Why does MiniMQTT need additional retries?

Reply:
Same comment as above. MiniMQTT has it's own built in retrying and some exceptions are retriable and others are fatal

Reply:
I think this code should be able to implement the same policy. I don't know why they error behavior would be specific to minimqtt.

Reply:
This logic is now removed